wire-ios-data-model/Source/MLS/OneOnOne/OneOnOneResolver.swift

```diff
@@ -92,68 +91,151 @@ public final class OneOnOneResolver: OneOnOneResolverInterface {
+        in context: NSManagedObjectContext
+    ) async -> OneOnOneConversationResolution {
+        WireLogger.conversation.debug("no common protocols found")
+
+        await context.perform {
+            guard
+                let otherUser = ZMUser.fetch(with: userID, in: context),
+                let conversation = otherUser.oneOnOneConversation
+            else {
+                return
+            }
 
-                conversation.isForcedReadOnly = true
+            self.makeConversationReadOnly(
+                selfUser: ZMUser.selfUser(in: context),
+                otherUser: otherUser,
+                conversation: conversation
+            )
+        }
 
-            }
-            return .archivedAsReadOnly
+        return .archivedAsReadOnly
+    }
 
-        case .mls:
-            WireLogger.conversation.debug("should resolve to mls 1-1 conversation")
+    private func makeConversationReadOnly(
+        selfUser: ZMUser,
+        otherUser: ZMUser,
+        conversation: ZMConversation
+    ) {
+        if conversation.isForcedReadOnly { return }
 
-            guard let migrator else {
-                throw OneOnOneResolverError.migratorNotFound
-            }
+        if !selfUser.supportedProtocols.contains(.mls) {
+            conversation.appendMLSMigrationMLSNotSupportedForSelfUser(user: selfUser)
+        } else if !otherUser.supportedProtocols.contains(.mls) {
+            conversation.appendMLSMigrationMLSNotSupportedForOtherUser(user: otherUser)
+        }
+
+        conversation.isForcedReadOnly = true
+    }
+
+    // MARK: Resolve - MLS
+
+    private func resolveCommonUserProtocolMLS(
+        with userID: QualifiedID,
+        in context: NSManagedObjectContext
+    ) async throws -> OneOnOneConversationResolution {
+        WireLogger.conversation.debug("should resolve to mls 1-1 conversation")
 
-            let mlsGroupIdentifier = try await migrator.migrateToMLS(
+        guard let migrator else {
+            throw OneOnOneResolverError.migratorNotFound
+        }
+
+        guard let mlsService else {
+            throw OneOnOneResolverError.migratorNotFound
+        }
+
+        let mlsGroupID = try await syncMLSConversationFromBackend(
+            userID: userID,
+            in: context
+        )
+
+        if await mlsService.conversationExists(groupID: mlsGroupID) {
+            return .noAction
+        }
+
+        guard let epoch = await fetchMLSConversationEpoch(mlsGroupID: mlsGroupID, in: context) else {
+            throw MigrateMLSOneOnOneConversationError.missingConversationEpoch
+        }
+
+        if epoch == 0 {
+            // migrate to a new conversation
+            try await migrator.migrateToMLS(
                 userID: userID,
+                mlsGroupID: mlsGroupID,
                 in: context
             )
+        } else {
+            // join existing conversation via external commit
+            try await mlsService.joinGroup(with: mlsGroupID)
```

Comment:
This is not enough we also need to perform what's done inside `migrator.migrateToMLS()` with the difference that we join the group instead of establishing it.

Reply:
What exactly? If we just join locally a 1-1 conversation, we have no history to copy

Reply:
See other discussion below